### PR TITLE
#889 Use linux style xdg on mac

### LINKED ISSUE #889: On macos: Respect XDG base dirs.
Apropos to #405, a large number of developer tools (especially those with a cli) respect XDG base dirs in both macos and linux. Those that insist on using `~/Library` on macos are outliers, as it creates unnecessary headaches for anyone that (a) maintains their own user configurations and / or (b) frequently switches between macos and linux. 

Can `sqlfluff` at least first look for configuration in `${XDG_CONFIG_HOME:-~/.config}` first on macos? Obviously `~/Library` can be used as a fallback if desired for some reason, though I think it becomes irrelevant for configuration / data if the XDG base dir spec is followed. (As indicated in https://stackoverflow.com/questions/3373948/equivalents-of-xdg-config-home-and-xdg-data-home-on-mac-os-x caching is a little more subtle).

### PR DESCRIPTION
**Pull request by dmateusp**
Resolves #889

### CODE REVIEW

File: src/sqlfluff/core/config.py [hunk 1]
```diff
@@ -260,14 +260,24 @@ def load_config_at_path(self, path: str) -> dict:
...
+    def load_user_appdir_config(self) -> Tuple[Optional[str], dict]:
         """Load the config from the user's OS specific appdir config directory."""
         appname = "sqlfluff"
         appauthor = "sqlfluff"
-        user_config_dir_path = appdirs.user_config_dir(appname, appauthor)
+
+        # On Mac OSX follow Linux XDG base dirs
+        # https://github.com/sqlfluff/sqlfluff/issues/889
+        user_config_dir_path = os.path.expanduser("~/.config/sqlfluff")
+        if appdirs.system == "darwin":
```
Comment (barrywhart) [inline src/sqlfluff/core/config.py:271]: This looks a little tricky, switching the OS. Is there a known issue with `appdirs` requiring this technique?

File: src/sqlfluff/core/config.py [hunk 2]
```diff
@@ -260,14 +260,24 @@ def load_config_at_path(self, path: str) -> dict:
...
+            appdirs.system = "linux2"
+            user_config_dir_path = appdirs.user_config_dir(appname, appauthor)
+            appdirs.system = "darwin"
+        if not os.path.exists(user_config_dir_path):
+            user_config_dir_path = appdirs.user_config_dir(appname, appauthor)
+
         if os.path.exists(user_config_dir_path):
-            return self.load_config_at_path(user_config_dir_path)
-        return {}
+            return user_config_dir_path, self.load_config_at_path(user_config_dir_path)
```
Comment (barrywhart) [inline src/sqlfluff/core/config.py:279]: It looks like the first return value exists solely so the automated test can check it. I wonder if we could refactor (split) this into two functions:
* One that returns the directory
* Another that calls the first and returns the config path

I think that may be cleaner, so we're not returning unneeded data to the "real" code. 🤔

Comment (dmateusp) [reply src/sqlfluff/core/config.py]: 100%, lazy on my end ^^

File: src/sqlfluff/core/config.py (hunk 1 above)
Comment (dmateusp) [reply src/sqlfluff/core/config.py:310]: the context is here: https://github.com/ActiveState/appdirs/issues/78#issuecomment-366947045

**Review by barrywhart — APPROVED** [review]
Looks good! I see the automated test. This seems like something we'd want to test manually as well. Have you done that?

**Comment by dmateusp** [comment]
Thanks for the suggestion @barrywhart! So I tested it manually like this:

``` bash
❯ git status
On branch dmateusp/#889/use_linux_style_xdg_on_mac
❯ export XDG_CONFIG_HOME=~/.config/linters
❯ echo "select 1 FROM table_a" > test.sql
❯ echo "[sqlfluff:rules:L010]\ncapitalisation_policy = upper" > ~/Library/Application\ Support/sqlfluff/.sqlfluff
❯ echo "[sqlfluff:rules:L010]\ncapitalisation_policy = lower" > ${XDG_CONFIG_HOME}/sqlfluff/.sqlfluff
❯ sqlfluff fix test.sql
❯ cat test.sql
select 1 from table_a
❯ git checkout master
Switched to branch 'master'
❯ sqlfluff fix test.sql
❯ cat test.sql
SELECT 1 FROM table_a
```

So I think it's working as expected

**Comment by dmateusp** [comment]
@barrywhart this made me realize that the error message in L010 isn't very helpful so I opened https://github.com/sqlfluff/sqlfluff/issues/1128

**Pull request merged**